contracts/datastore/DataStore.sol

```diff
@@ -0,0 +1,362 @@
+pragma solidity ^0.5.0;
+
+import "../interfaces/ISecurityToken.sol";
+import "../interfaces/IOwnable.sol";
+import "../interfaces/IDataStore.sol";
+import "./DataStoreStorage.sol";
+
+/**
+ * @title Data store contract that stores data for all the modules in a central contract.
+ */
+contract DataStore is DataStoreStorage, IDataStore {
+    //NB To modify a specific element of an array, First push a new element to the array and then delete the old element.
+    //Whenver an element is deleted from an Array, last element of that array is moved to the index of deleted element.
+    //Delegate with MANAGEDATA permission can modify data.
+
+    modifier onlyAuthorized() {
+        bool isOwner = msg.sender == IOwnable(address(securityToken)).owner();
+        require(isOwner || 
+            securityToken.isModule(msg.sender, DATA_KEY) || 
+            securityToken.checkPermission(msg.sender, address(this), MANAGEDATA), 
+            "Unauthorized"
+        );
+        _;
+    }
+
+    modifier validKey(bytes32 _key) {
+        require(_key != bytes32(0), "Missing key");
+        _;
+    }
+
+    modifier validArrayLength(uint256 _keyLength, uint256 _dataLength) {
+        require(_keyLength == _dataLength, "Array length mismatch");
+        _;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == IOwnable(address(securityToken)).owner(), "Unauthorized");
+        _;
+    }
+
+    /**
+     * @dev Changes security token atatched to this data store
+     * @param _securityToken address of the security token
+     */
+    function setSecurityToken(address _securityToken) external onlyOwner {
+        if(address(securityToken) != address(0)) {
+            require(msg.sender == IOwnable(address(securityToken)).owner(), "Unauthorized");
```

Comment:
Isn't this already done through the `onlyOwner` modifier?

Reply:
Will modify to just have the modifier (which is needed to stop someone changing the security token to address(0) which would break things).

Reply:
Thinking a bit more about this, this seems a slightly dangerous operation. An issuer could assign the security token address to be anything (i.e. not an actual security token) which means that the data could then be updated adhoc. The original ST though would still think the data store belonged to it.

I do think it is useful to be able to assign a dataStore to a new token (so that if an issuer wants to migrate its data to a new token it can do) but wondering whether we should also:
  - 1. check the new address is a ST (using the STR)
  - 2. tell the original underlying token that it no longer owns the dataStore

Reply:
If we do 1), we won't need 3). If gas limit allows, I think we should add the check just to minimise mistaken address changes.

2) I don't think we should prevent or worry too much about this. Issuer has a lot more power anyway. In case of upgrading token, Keeping the data store active for both tokens (read for old, write for new) can be useful to create a bridge that will help people migrate their security tokens.

Reply:
OK - for now I've just made it so that the function has an onlyOwner modifier. As you say, for initialisation it is done through the constructor, so any subsequent changes should only be made by the ST owner. I'll add checking in the STR as a subsequent PR - not too worried about gas costs as it will be a rare operation.

Reply:
I meant to say code size limit and not gas limit :)